test/XmlHttpRequestConnection.tests.js

```diff
@@ -145,6 +145,80 @@ define(function (require) {
 
         });
 
+        describe("is correctly using XmlHttpRequest and runs attached callbacks while performing:", function (){
+            beforeEach(function (){
+                mockXMLHttpRequest.prototype.send = function (body) {
+                    this.readyState = 4;
+                    this.status = 200;
+                    this.onloadstart();
+                    this.onload();
+                    this.onloadend();
+                    this.onprogress();
+                    this.ontimeout();
+                    this.onerror();
+                    this.onabort();
+                    this.upload.onloadstart();
+                    this.upload.onload();
+                    this.upload.onloadend();
+                    this.upload.onprogress();
+                    this.upload.ontimeout();
+                    this.upload.onerror();
+                    this.upload.onabort();
+                    this.onreadystatechange();
+                };
+                mockXMLHttpRequest.prototype.upload = {};
+                spyOn(mockXMLHttpRequest.prototype, 'send').andCallThrough();
+
+                window.XMLHttpRequest = (function () { return mockXMLHttpRequest; }());
+
+                connection = new XmlHttpRequestConnection();
+            });
+
+            it("execute call and invokes custom request event callbacks", function () {
```

Comment:
this test tests too many things at the same time, there should one test per event handler. It could also be simplified a bit because here you are kind of reimplementing XmlHttpRequest event system while what you should test is that the event handler is registered (note: using `addEventListener` as suggested in a previous comment would have made that a lot easier). So I would write, something like:

```js
describe('register XmlHttpRequest event handler', function (){
    beforeEach(function (){
        mockXMLHttpRequest.prototype.upload = {};
        window.XMLHttpRequest = (function () { return mockXMLHttpRequest; }());

        connection = new XmlHttpRequestConnection();
    });

    function testEventHandler(eventHandlerName) {
        var eventHandler = function () {},
            events = {};

        events[eventHandlerName] = eventHandler;

        mockXMLHttpRequest.prototype.send = function () {
            expect(this[eventHandlerName]).toBe(eventHandler);
        };
        connection.execute(mockRequest, events, mockCallback);
    }

    function testUploadEventHandler(eventHandlerName) {
        var eventHandler = function () {},
            events = {'upload': {}};

        events.upload[eventHandlerName] = eventHandler;

        mockXMLHttpRequest.prototype.send = function () {
            expect(this.upload[eventHandlerName]).toBe(eventHandler);
        };
        connection.execute(mockRequest, events, mockCallback);
    }

    it('should register onloadstart event handler', function () {
        testEventHandler('onloadstart');
    });

    // other event handlers...

    it('should register upload onloadstart event handler', function () {
        testUploadEventHandler('onloadstart');
    });

    // others upload events handlers...
});
```

and actually you should also test that for instance a `onwhateverevent` is completely ignored.